.changeset/soft-berries-obey.md

```diff
@@ -0,0 +1,8 @@
+---
+'@leafygreen-ui/hooks': minor
+---
+
+- Extends `useControlledValue` to accept any type.
+- Adds `updateValue` function in return value. This method triggers a synthetic event to update the value of a controlled or uncontrolled component.
```

Comment:
Im a little confused by this. Can you explain what this is solving?

Reply:
If a component is using this hook, there's no easy way to programmatically change the value.

Consider the component below. If we want to update the component's value (or trigger a change in the consuming component) on button click, we would need to call `handleChange` and pass in a synthetic change event. The `updateValue` function makes it cleaner to do this.

``` tsx
    const TestComponent = ({ valueProp, handlerProp }) => {
      const inputRef = useRef<HTMLInputElement>(null);
      const { value, handleChange, updateValue } = useControlledValue(
        valueProp,
        handlerProp,
      );

      return (
        <>
          <input
            ref={inputRef}
            value={value}
            onChange={handleChange}
          />
          <button
            onClick={() => updateValue('carrot', inputRef)}
          />
        </>
      );
    };
```

Note: `updateValue` takes an element ref, since we still need to simulate a change event under the hood, and we need to know what element should be the trigger

Reply:
One update I'm considering making is to pass the element ref directly into the hook, so the call would be something like

```tsx
const inputRef = useRef<HTMLInputElement>(null);
const { value, handleChange, updateValue } = useControlledValue(
  valueProp,
  handlerProp,
  inputRef
);

...
<button
  onClick={() => updateValue('carrot')}
/>
```

Reply:
Thanks for explaining! I like the idea of passing the ref directly to the hook.

Reply:
But there could be confusion as to why `updateValue` won't work if `inputRef` is optional and not passed to the hook.